File: MarsRoverTechChallenge/OO/Parser.cs
using System.Collections.Generic;
using System.Linq;
using MarsRoverTechChallenge.Shared;

namespace MarsRoverTechChallenge.OO
{
    public class Parser
    {
        public static Input Parse(string input)
        {
            var lines = input.Split('\n').Select(s => s.Trim()).ToArray();
            var fieldSize = ParseFieldSize(lines[0]);
            var instructions = new List<RoverInstructions>();
            for (var i = 1; i < lines.Length; i += 2)
                instructions.Add(ParseInstructions(lines[i], lines[i + 1]));

            return new Input(fieldSize, instructions);
        }

        private static Point ParseFieldSize(string line)
        {
            var split = line.Split(' ');
            return new Point(int.Parse(split[0]), int.Parse(split[1]));
        }

        private static readonly Dictionary<char, Direction> _directionMap = new Dictionary<char, Direction>()
        {
            {'N', Direction.North},
            {'E', Direction.East},
            {'S', Direction.South},
            {'W', Direction.West},
        };


        private static readonly Dictionary<char, Movement> _movementMap = new Dictionary<char, Movement>()
        {
            {'L', Movement.Left},
            {'R', Movement.Right},
            {'M', Movement.Forward},
        };

        private static RoverInstructions ParseInstructions(string startLine, string movement)
        {
            var split = startLine.Split(' ');
            var position = new Position(int.Parse(split[0]), int.Parse(split[1]), _directionMap[split[2][0]]);
            var instructions = movement.Select(c => _movementMap[c]).ToArray();
            return new RoverInstructions(position, instructions);
        }
    }
}

File: FSharpRover/Parser.fs
module Parser

open System
open Types

let parse (input:string) = 
    let (|Int|_|) str =
        match System.Int32.TryParse(str) with
        | (true,int) -> Some(int)
        | _ -> None

    let (|Direction|_|) d = 
        match d with
        | "N" -> Some North
        | "E" -> Some East
        | "S" -> Some South
        | "W" -> Some West
        | _ -> None

    let (|Movement|_|) m = 
        match m with
        | 'L' -> Some Left
        | 'R' -> Some Right
        | 'M' -> Some Forward
        | _ -> None

    let rec (|MovementsC|_|) chars = 
        match chars with
        | Movement m :: MovementsC tail -> Some (m :: tail)
        | [] -> Some []
        | _ -> None

    let (|Movements|_|) (str:string) = 
        match List.ofArray (str.ToCharArray()) with
        | MovementsC m -> Some m
        | _ -> None

    let (|BoardSize|_|) (str:string) = 
        match str.Split ' ' with
        | [|Int x; Int y|] -> Some (x, y)
        | _ ->  None
    
    let (|Position|_|) (str:string) =
        match str.Split ' ' with
        | [|Int x; Int y; Direction d|] -> Some (x, y, d)
        | _ -> None

    let rec (|RobotInstructions|_|) lines = 
        match lines with
        | Position p :: Movements m :: RobotInstructions tail -> Some ((p, m) :: tail)
        | [] -> Some []
        | _ -> None

    let splitLines (str:string) = 
        let trim (s:string) = s.Trim();
        (input.Split '\r') |> List.ofArray |> List.map trim

    match splitLines input with
        | BoardSize l :: RobotInstructions i -> (l,i)
        | _ -> failwith "Invalid Input" 